applications/luci-app-speedtest/htdocs/luci-static/resources/view/speedtest.js

```diff
@@ -0,0 +1,141 @@
+				E('option', { 'value': 'latency' }, _('Latency')),
+				E('option', { 'value': 'download' }, _('Download')),
+				E('option', { 'value': 'upload' }, _('Upload'))
+			])
+		];
+		var status = [
+			E('label', { 'class': 'cbi-input-label', 'style': 'margin-right: 8px;'}, _('Status')),
+			E('em', { 'id': 'speedtest-status'}, _('Available'))
+		];
+		var table = [
+			E('h3', {'class': 'section-title'}, _('Results')),
+			E('table', {'class': 'table cbi-section-table'}, [
+				E('tr', {'class': 'tr table-title'}, [
+					E('th', {'class': 'th ', 'style': 'display: none;'})
+				]),
+				E('tr', {'class': 'tr cbi-rowstyle-1'},[
+					E('td', {'class': 'td left', 'width': '50%'}, _('IP Address')),
+					E('td', {'class': 'td', 'id': 'client-ip'}, '-')
+				]),
+				E('tr', {'class': 'tr cbi-rowstyle-2'},[
+					E('td', {'class': 'td left', 'width': '50%'}, _('Provider')),
+					E('td', {'class': 'td', 'id': 'client-isp'}, '-')
+				]),
+				E('tr', {'class': 'tr cbi-rowstyle-1'},[
+					E('td', {'class': 'td left', 'width': '50%'}, _('Server')),
+					E('td', {'class': 'td', 'id': 'server-name'}, '-')
+				]),
+				E('tr', {'class': 'tr cbi-rowstyle-2'},[
+					E('td', {'class': 'td left', 'width': '50%'}, _('ID')),
+					E('td', {'class': 'td', 'id': 'server-id'}, '-')
+				]),
+				E('tr', {'class': 'tr cbi-rowstyle-1'},[
+					E('td', {'class': 'td left', 'width': '50%'}, _('Location')),
+					E('td', {'class': 'td', 'id': 'server-location'}, '-')
+				]),
+				E('tr', {'class': 'tr cbi-rowstyle-2'},[
+					E('td', {'class': 'td left', 'width': '50%'}, _('Server Host')),
+					E('td', {'class': 'td', 'id': 'server-host'}, '-')
+				]),
+				E('tr', {'class': 'tr cbi-rowstyle-1'},[
+					E('td', {'class': 'td left', 'width': '50%'}, _('Ping')),
+					E('td', {'class': 'td', 'id': 'ping'}, '-')
+				]),
+				E('tr', {'class': 'tr cbi-rowstyle-2'},[
+					E('td', {'class': 'td left', 'width': '50%'}, _('Jitter')),
+					E('td', {'class': 'td', 'id': 'jitter'}, '-')
+				]),
+				E('tr', {'class': 'tr cbi-rowstyle-1'},[
+					E('td', {'class': 'td left', 'width': '50%'}, _('Download Speed')),
+					E('td', {'class': 'td', 'id': 'download'}, '-')
+				]),
+				E('tr', {'class': 'tr cbi-rowstyle-2'},[
+					E('td', {'class': 'td left', 'width': '50%'}, _('Upload Speed')),
+					E('td', {'class': 'td', 'id': 'upload'}, '-')
+				])
+			])
+		];
+		var running = false;
+		var button = [
+			E('button', {
+				'class': 'btn cbi-button cbi-button-action',
+				'click': function() {
+					if (!running) {
+						running = true;
+						var dropdown = document.getElementById('speedtest-type');
+						var value = dropdown.options[dropdown.selectedIndex].value;
+						var command = ['--output', 'json'];
+						if (value !== 'all') {
+							command.push('--' + value);
+						}
+
+						var status = document.getElementById('speedtest-status');
+						status.textContent = _('Running');
+
+						fs.exec_direct('speedtest', command)
```

Comment:
So the app assumes that the speedtest binary is available. Let's say that the binary was deleted, or some sysupgrade went bad (even tho this app depends on it), what happens when the binary is absent? Is the error message misleading or must we perform a separate check at init?

Reply:
The action has been handled here.
```js
.catch(function(err) {
	if (!err.response) {
		status.textContent = _('No response received, please check your internet connection or try again later.');
	} else if (err.response.data && err.response.data.error === 'unable to retrieve your ip info') {
		status.textContent = _('Unable to retrieve IP information, please check your internet connection.');
	} else {
		status.textContent = err;
	}
	running = false;
});
```